clippy_lints/src/exit.rs

```diff
@@ -0,0 +1,62 @@
+use crate::utils::{match_def_path, paths, qpath_res, span_lint};
+use if_chain::if_chain;
+use rustc::hir::{Expr, ExprKind, Item, ItemKind, Node};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** `exit()`  terminates the program and doesn't provide a
+    /// stack trace.
+    ///
+    /// **Why is this bad?** Ideally a program is terminated by finishing
+    /// the main function.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// std::process::exit(0)
+    /// ```
+    pub EXIT,
+    restriction,
+    "`std::process::exit` is called, terminating the program"
+}
+
+declare_lint_pass!(Exit => [EXIT]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Exit {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, e: &'tcx Expr) {
+        if_chain! {
+            if let ExprKind::Call(ref path_expr, ref _args) = e.kind;
+            if let ExprKind::Path(ref path) = path_expr.kind;
+            if let Some(def_id) = qpath_res(cx, path, path_expr.hir_id).opt_def_id();
+            if match_def_path(cx, def_id, &paths::EXIT);
+            then {
+                let mut parent = cx.tcx.hir().get_parent_item(e.hir_id);
+                // We have to traverse the parents upwards until we find a function
+                // otherwise a exit in a let or if in main would still trigger this
+                loop{
+                    match cx.tcx.hir().find(parent) {
+                        Some(Node::Item(Item{ident, kind: ItemKind::Fn(..), ..})) => {
+                            // If we found a function we check it's name if it is
+                            // `main` we emit a lint.
+                            if ident.name.as_str() != "main" {
```

Comment:
maybe replace this check with `utils::is_entrypoint_fn`

Reply:
updated